File: 2021/day-02.py
data = [[com, int(x)] for (com, x) in
        [line.strip().split(' ') for line in open('input/day-02-input.txt')]]

def solve_1(data):
    hor = dep = 0
    for com, x in data:
        if com == 'forward':
            hor += x
        elif com == 'down':
            dep += x
        elif com == 'up':
            dep -= x
    return hor * dep

def solve_2(data):
    hor = dep = aim = 0
    for com, x in data:
        if com == 'forward':
            hor += x
            dep += aim * x
        elif com == 'down':
            aim += x
        elif com == 'up':
            aim -= x
    return hor * dep

print(solve_1(data)) # 1762050
print(solve_2(data)) # 1855892637

File: 2015/erlang/day_02.erl
-module(day_02).
-export([main/0]).

main() ->
    Input = readlines("../input/day-02.txt"),
    {TotalArea, TotalRibbon} =
        lists:foldl(
            fun(Line, {AreaAcc, RibbonAcc}) ->
                Dimensions = parse_dimensions(Line),
                {AreaAcc + calculate_area(Dimensions),
                 RibbonAcc + calculate_ribbon(Dimensions)}
            end,
            {0, 0},
            Input
        ),
    io:fwrite("Area: ~p, Ribbon: ~p~n", [TotalArea, TotalRibbon]).

readlines(FileName) ->
    {ok, Data} = file:read_file(FileName),
    binary:split(Data, [~"\r\n", ~"\n"], [global, trim_all]).

parse_dimensions(Line) ->
    [binary_to_integer(X) || X <- binary:split(Line, ~"x", [global])].

calculate_area([L, W, H]) ->
    Sides = [L * W, W * H, H * L],
    2 * lists:sum(Sides) + lists:min(Sides).

calculate_ribbon(Dimensions) ->
    M = lists:sum(Dimensions) - lists:min(Dimensions),
    2 * M + lists:foldl(fun(X, Prod) -> X * Prod end, 1, Dimensions).
